### ISSUE TITLE
Simplify and split apart MakePod

### ISSUE THREAD

**Issue by imjasonh** [tektoncd/pipeline#1605]
Related to #1519

Today, the conversion between TaskRun to a Pod jumbled together (mostly) inside [`MakePod`](https://github.com/tektoncd/pipeline/blob/c9a7a35c01aea3ad6cf8b0992e10b0a21535d2a6/pkg/reconciler/taskrun/resources/pod.go#L112-L292), which is responsible for lots of different things. This makes it hard to understand, hard to test in isolation, and hard to make fixes and improvements.

## What `MakePod` does

(roughly in the ideal order)

### Applies `stepTemplate`
* FIX: This should ignore `stepTemplate.command` and `.args` if the step specifies `.script`
* FIX: `stepTemplate` shouldn't apply to `sidecars`, only `steps`.

### Convert Steps with `script` to Containers with `command`
* This adds an init container, `place-scripts`, which writes the user's scripts to `/builder/scripts`
* This adds a new volume, which needs to be mounted into all steps that use a script.
* Done in #1616

### Resolve image `ENTRYPOINT` from remote registry if the step doesnt' specify a `command`
* This relies on [`k8schain`](https://godoc.org/github.com/google/go-containerregistry/pkg/authn/k8schain)
* FIX: Also resolve the image's digest, to avoid a race between entrypoint resolving and container running
* FIX: Cache tag -> digest lookup for each resolve, for speed
* Done in #1616

### Rewrite step `command` + `args` to inject our entrypoint binary
* Adds an init container, `place-tools`, always.
* Adds a volume, `/builder/tools`
* Adds a downward API volume at `/builder/downward`
* Entrypoint rewriting should happen _after_ `script` conversion (today it happens before)
* Done in #1616

### Initialize credentials
* Adds an init container, `creds-init`, if annotated creds are found in the service account's secrets
* Adds `/var/secrets/...` for each secret
* Requires `/builder/home` volume and (I think) `HOME` env var
* Split out in https://github.com/tektoncd/pipeline/pull/1599

### `mkdir` working dirs under `/workspace`
* Adds an init container, `working-dir-init`, if any steps specify a relative `workingDir`
* Split out in https://github.com/tektoncd/pipeline/pull/1596

### Apply resource requests/limits
* Find the maximum request per resource per step
* Apply that request to step 0
* Set all other steps' requests to 0, and limits to the max request.

### Other minor operations
* Set `workingDir` to `/workspace` if the user didn't specify
* Set the container `name` to `step-%s` (step name), or to `step-unnamed-%d` (step index) if not specified

## After these changes
* We can more easily make entrypoint changes described in https://github.com/tektoncd/pipeline/issues/1569
* Generate any volume names to avoid conflict
* Rename `/builder/` volumeMount paths to `/tekton/`

**Comment by vdemeester** [comment]
@ImJasonH I wasn't sure if what is listed is what `MakePod` does currently (and that should change somehow) or what it should do ?

/kind feature

**Comment by ImJasonH** [comment]
This lists what MakePod does, in the ~ideal order. Before these changes things happened in jumbled order because it mostly happened in one huge wall of text. Afterwards I'd like to have one `MakePod` method that just calls N smaller isolated methods with clear goals and tests.

**Comment by afrittoli** [comment]
Today the `stepTemplate` is also applied to injected steps like those from `PipelineResources`. 
I think that may be ok, but we should probably document it.

**Comment by ImJasonH** [comment]
Yeah I _think_ that's expected behavior -- at least, I think that was the behavior before I started reorganizing `MakePod` -- but I agree it should be better documented.

**Comment by ImJasonH** [comment]
After https://github.com/tektoncd/pipeline/pull/1655 I think I'm comfortable declaring victory on this.

We might want to move PipelineResource interactions into MakePod as well (prepending/appending steps for resources, applying parameter replacements), but at least those are already fairly well isolated into their own methods with clear tests.

**Comment by lbernick** [comment]
/reopen

There are still a few TODOs associated with this issue-- this came up during review of #4598.

It's not clear where mergeSteps should live; the TODO [here](https://github.com/tektoncd/pipeline/blob/38d98f09ecdd5240d91762ff6a75bc6275316ed9/pkg/pod/pod.go#L136) would result in a circular import as this functionality is used in API validation which is a bit weird. @wlynch suggested moving into an [internal package](https://go.dev/doc/go1.4#internalpackages).

**Comment by vdemeester** [comment]
Closing this issue — the `MakePod` function has been fully decomposed over the years through multiple PRs (#1616, #1621, #1623, #1627, #1644, #1655, #1686, #4598, and others). The monolithic function no longer exists in the codebase, and the concerns it handled (stepTemplate application, script conversion, entrypoint rewriting, credential initialization, working dir init, resource requests) are now handled by separate, well-tested components.

All the sub-items listed in the original issue have been addressed. 🎉
